include/rocksdb/iterator.h

```diff
@@ -117,6 +117,46 @@ class Iterator : public Cleanable {
     assert(false);
     return Slice();
   }
+
+  /**
+   * true if this iterator is already
+   * checking ReadOptions::iterate_lower_bound.
+   * false indicates that either the iterator does
+   * not check the lower bound, or that it does not
+   * report that it is performing the check.
+   */
+  virtual bool ChecksLowerBound() const = 0;
+
+  /**
+   * Returns the lower bound if this iterator has a
+   * ReadOptions::iterate_lower_bound set, else
+   * returns nullptr.
+   *
+   * Just because a lower bound is present, it does
+   * not mean that it is checked. This can however
+   * be determined by calling Iterator::ChecksLowerBound().
+   */
+  virtual const Slice* lower_bound() const = 0;
+
+  /**
+   * true if this iterator is already
+   * checking ReadOptions::iterate_upper_bound.
+   * false indicates that either the iterator does
+   * not check the upper bound, or that it does not
+   * report that it is performing the check.
+   */
+  virtual bool ChecksUpperBound() const = 0;
```

Comment:
Can you explain it? RocksDB's Iterator is always supposed to have upper bound checked. Did you see a bug somewhere?

Reply:
I didn't find that to be the case. There are iterators which don't check upper/lower bounds, which is what triggered this PR. When you wrap an iterator, if the wrapped iterator does not implement upper/lower bounds checks, you may need to know that, so that you can implement the checks in your wrapper iterator to get the behaviour you need.